### PR TITLE
add `TargetTrackingScaling` as a scaling policy option

### PR DESCRIPTION
**Pull request by falcon78921**
##### SUMMARY

Add `TargetTrackingScaling` as an EC2 scaling option.
Fixes: https://github.com/ansible-collections/community.aws/issues/544

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ec2_scaling_policy.py

### CODE REVIEW

**Comment by falcon78921** [comment]
As it is now, `predefined_metric_spec` seems to work:

~~~
changed: [127.0.0.1] => {
    "alarms": [
        {
...
...
...
    "as_name": "asg-test-1",
    "auto_scaling_group_name": "asg-test-1",
    "changed": true,
    "enabled": true,
    "invocation": {
        "module_args": {
            "adjustment_type": null,
            "asg_name": "asg-test-1",
            "aws_access_key": null,
            "aws_ca_bundle": null,
            "aws_config": null,
            "aws_secret_key": null,
            "cooldown": null,
            "debug_botocore_endpoint_logs": false,
            "ec2_url": null,
            "estimated_instance_warmup": null,
            "metric_aggregation": "Average",
            "min_adjustment_step": null,
            "name": "test123",
            "policy_type": "TargetTrackingScaling",
            "profile": null,
            "region": null,
            "scaling_adjustment": null,
            "security_token": null,
            "state": "present",
            "step_adjustments": null,
            "target_tracking_config": {
                "customized_metric_spec": null,
                "predefined_metric_spec": {
                    "predefined_metric_type": "ASGAverageCPUUtilization"
                },
                "target_value": 125.5
            },
            "validate_certs": true
        }
    },
    "name": "test123",
...
...
...
    "policy_name": "test123",
    "policy_type": "TargetTrackingScaling",
    "step_adjustments": [],
    "target_tracking_configuration": {
        "disable_scale_in": false,
        "predefined_metric_specification": {
            "predefined_metric_type": "ASGAverageCPUUtilization"
        },
        "target_value": 125.5
    }
}
~~~

File: plugins/modules/ec2_scaling_policy.py [hunk 1]
```diff
@@ -355,28 +531,54 @@ def main():
     step_adjustment_spec = dict(
         lower_bound=dict(type='int'),
         upper_bound=dict(type='int'),
-        scaling_adjustment=dict(type='int', required=True))
+        scaling_adjustment=dict(type='int', required=True)
+    )
+
+    # Spec for TargetTrackingScaling policies
+    target_tracking_spec = dict(
```
Comment (jillr) [inline plugins/modules/ec2_scaling_policy.py:538]: You're missing some types in this spec.  I find it cleaner to break out the options into separate vars as this example, it's a stylistic preference however you prefer it. 
```
    # Specs for TargetTrackingScaling policies
    predefined_metric_spec = dict(
        predefined_metric_type=dict(type='str', choices=['ASGAverageCPUUtilization',
                                                         'ASGAverageNetworkIn',
                                                         'ASGAverageNetworkOut',
                                                         'ALBRequestCountPerTarget']),
        resource_label=dict(type='str')
    )
    customized_metric_spec = dict(
        metric_name=dict(type='str', required=True),
        namespace=dict(type='str', required=True),
        statistic=dict(type='str', required=True, choices=['Average', 'Minimum', 'Maximum', 'SampleCount', 'Sum']),
        dimensions=dict(type='list', elements='str'),
        unit=dict(type='str', required=True)
    )
```
Then below in the arg_spec:
```
        target_tracking_spec=dict(options=dict(
            disable_scalein=dict(type='bool'),
            target_value=dict(type='float', required=True),
            predefined_metric_spec=dict(type='dict',
                                        options=predefined_metric_spec),
            customized_metric_spec=dict(type='dict',
                                        options=customized_metric_spec)
        )),
```

**Review by jillr — COMMENTED** [review]
Good progress on this, thanks @falcon78921!

File: plugins/modules/ec2_scaling_policy.py [hunk 2]
```diff
@@ -267,18 +337,113 @@ def create_scaling_policy(connection, module):
...
+                msg='Only one metric specification can be used at a time with target_tracking_config')
+        elif target_tracking_config['predefined_metric_spec']:
+            targetTrackingConfig['PredefinedMetricSpecification'] = dict()
+        elif target_tracking_config['customized_metric_spec']:
+            targetTrackingConfig['CustomizedMetricSpecification'] = dict()
+
+        if module.params['target_tracking_config']['predefined_metric_spec']:
+            # Boto3 documentation says that AdjustmentType is only required for SimpleScaling
+            # and StepScaling, not for TargetTrackingScaling. However, Boto3 fails on validation
+            # if a NoneType is passed. If TargetTrackingScaling is chosen and adjustment_type is None,
```
[the remark this answers is not in the document]
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py:365]: @jillr I can try it and see. The boto3 docs mention `AdjustmentType` is required if policy is `StepScaling` or `SimpleScaling` (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/autoscaling.html#AutoScaling.Client.put_scaling_policy).

[the remark this answers is not in the document]
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py]: @jillr Ok, so boto3 complains of a `NoneType` value for `AdjustmentType` even if the policy is `TargetTrackingScaling`. According to the docs, only `SimpleScaling` and `StepScaling` require an `AdjustmentType`. In order to circumvent, I added a condition to send an explicit string with `None` if no `AdjustmentType` is provided for `TargetTrackingScaling`.

File: plugins/modules/ec2_scaling_policy.py [hunk 3]
```diff
@@ -267,18 +337,113 @@ def create_scaling_policy(connection, module):
...
+            # #AutoScaling.Client.put_scaling_policy
+            if target_tracking_config.get('adjustment_type') is None:
+                params['AdjustmentType'] = "None"
+
+            # Validate parameters for predefined_metric_spec
+            if target_tracking_config.get('target_value') is None:
+                module.fail_json(msg='target_value is required when policy_type is'
+                                     'TargetTrackingScaling, a predefined_metric_spec is defined, and state is present')
+            else:
+                targetTrackingConfig['TargetValue'] = target_tracking_config['target_value']
```
[the remark this answers is not in the document]
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py]: I think this should be resolved with the newly added `build_target_specification()`

File: plugins/modules/ec2_scaling_policy.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py]: Ok, I think `AdjustmentType` defaulted to `None` should work.

File: plugins/modules/ec2_scaling_policy.py [hunk 4]
```diff
@@ -267,18 +337,113 @@ def create_scaling_policy(connection, module):
...
+
+        if module.params['target_tracking_config']['predefined_metric_spec']:
+            # Boto3 documentation says that AdjustmentType is only required for SimpleScaling
+            # and StepScaling, not for TargetTrackingScaling. However, Boto3 fails on validation
+            # if a NoneType is passed. If TargetTrackingScaling is chosen and adjustment_type is None,
+            # pass None as a string
+            # See: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/autoscaling.html
+            # #AutoScaling.Client.put_scaling_policy
+            if target_tracking_config.get('adjustment_type') is None:
+                params['AdjustmentType'] = "None"
```
[the remark this answers is not in the document]
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py]: Added a default of `None` to the argument spec.

File: plugins/modules/ec2_scaling_policy.py [hunk 5]
```diff
@@ -267,18 +433,19 @@ def create_scaling_policy(connection, module):
...
         if module.params['cooldown']:
             params['Cooldown'] = module.params['cooldown']
 
-    if policy_type == 'StepScaling':
+    elif policy_type == 'StepScaling':
         if not module.params['step_adjustments']:
-            module.fail_json(msg='step_adjustments is required when policy_type is StepScaling '
-                             'and state is present')
+            module.fail_json(msg='step_adjustments is required when policy_type is StepScaling'
+                          'and state is present')
```
Comment (jillr) [inline plugins/modules/ec2_scaling_policy.py:444]: it looks like your editor has "format on save" enabled, this creates a lot of unnecessary diff noise which can be difficult to review (now, and if we need to go through git history later for bugfixes).  would you be able to toggle this setting off for this repo and back out the formatting changes in this file please?

File: plugins/modules/ec2_scaling_policy.py [hunk 6]
```diff
@@ -303,7 +479,6 @@ def create_scaling_policy(connection, module):
         policy = policies[0]
         for key in params:
             if params[key] != policy.get(key):
-                changed = True
```
Comment (jillr) [inline plugins/modules/ec2_scaling_policy.py:306]: We should still need this `changed` var to be set here.

Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py]: I think we're good to go here: https://github.com/ansible-collections/community.aws/blob/b5a7e2388488d347d6d6ff8af6e23b88fcb5af69/plugins/modules/ec2_scaling_policy.py#L480

**Comment by falcon78921** [comment]
ansible-test-units-community-aws-python38 fails here:

```
2022-02-01 14:55:14.762412 | controller | [gw0] [ 97%] FAILED tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_check_for_update_tags
2022-02-01 14:55:15.142097 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_check_for_update_nonmodifiable_attr
2022-02-01 14:55:15.143054 | controller | [gw0] [ 98%] FAILED tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_check_for_update_nonmodifiable_attr
2022-02-01 14:55:15.529054 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_add_tags
2022-02-01 14:55:15.530078 | controller | [gw0] [ 99%] FAILED tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_add_tags
2022-02-01 14:55:16.143226 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_remove_tags
2022-02-01 14:55:16.224015 | controller | [gw0] [100%] FAILED tests/unit/plugins/modules/test_ec2_vpc_vpn.py::test_remove_tags
2022-02-01 14:55:16.224050 | controller |
2022-02-01 14:55:16.224054 | controller | =================================== FAILURES ===================================
2022-02-01 14:55:16.224058 | controller | _______________________________ test_upload_api ________________________________
2022-02-01 14:55:16.224061 | controller | [gw0] linux -- Python 3.8.12 /home/zuul/venv/bin/python
2022-02-01 14:55:16.224064 | controller | monkeypatch = <_pytest.monkeypatch.MonkeyPatch object at 0x7ff8e611bf70>
2022-02-01 14:55:16.224068 | controller |
2022-02-01 14:55:16.224071 | controller |     def test_upload_api(monkeypatch):
2022-02-01 14:55:16.224074 | controller |         class FakeConnection:
2022-02-01 14:55:16.224077 | controller |
2022-02-01 14:55:16.224080 | controller |             def put_rest_api(self, *args, **kwargs):
2022-02-01 14:55:16.224083 | controller |                 assert kwargs["body"] == "the-swagger-text-is-fake"
2022-02-01 14:55:16.224086 | controller |                 return {"msg": "success!"}
2022-02-01 14:55:16.224090 | controller |
2022-02-01 14:55:16.224093 | controller |         def return_fake_connection(*args, **kwargs):
2022-02-01 14:55:16.224096 | controller |             return FakeConnection()
2022-02-01 14:55:16.224099 | controller |
2022-02-01 14:55:16.224102 | controller | >       monkeypatch.setattr(core, "boto3_conn", return_fake_connection)
2022-02-01 14:55:16.224106 | controller | E       AttributeError: <module 'ansible_collections.amazon.aws.plugins.module_utils.core' (<ansible.utils.collection_loader.AnsibleCollectionLoader object at 0x7ff8ec5fd970>)> has no attribute 'boto3_conn'
2022-02-01 14:55:16.224110 | controller |
2022-02-01 14:55:16.224113 | controller | tests/unit/plugins/modules/test_aws_api_gateway.py:44: AttributeError
2022-02-01 14:55:16.224116 | controller | _________________________ test_find_connection_filters _________________________
2022-02-01 14:55:16.224119 | controller | [gw0] linux -- Python 3.8.12 /home/zuul/venv/bin/python
2022-02-01 14:55:16.224122 | controller | placeboify = Session(region_name='us-west-2'), maybe_sleep = None
2022-02-01 14:55:16.224125 | controller |
2022-02-01 14:55:16.224133 | controller |     def test_find_connection_filters(placeboify, maybe_sleep):
2022-02-01 14:55:16.224137 | controller |         # setup dependencies for 2 vpn connections
2022-02-01 14:55:16.224140 | controller | >       dependencies = setup_req(placeboify, 2)
2022-02-01 14:55:16.224143 | controller |
2022-02-01 14:55:16.224146 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py:139:
2022-02-01 14:55:16.224149 | controller | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2022-02-01 14:55:16.224152 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py:361: in setup_req
2022-02-01 14:55:16.224155 | controller |     m, conn = setup_mod_conn(placeboify, params)
2022-02-01 14:55:16.224158 | controller | _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
2022-02-01 14:55:16.224161 | controller |
2022-02-01 14:55:16.224164 | controller | placeboify = Session(region_name='us-west-2')
2022-02-01 14:55:16.224167 | controller | params = {'connection_type': 'ipsec.1', 'customer_gateway_id': 'cgw-6113c87f', 'delay': 15, 'filters': {}, ...}
2022-02-01 14:55:16.224171 | controller |
2022-02-01 14:55:16.224174 | controller |     def setup_mod_conn(placeboify, params):
2022-02-01 14:55:16.224177 | controller |         conn = placeboify.client('ec2')
2022-02-01 14:55:16.224180 | controller |         retry_decorator = aws_ec2.AWSRetry.jittered_backoff()
2022-02-01 14:55:16.224183 | controller | >       wrapped_conn = aws_core._RetryingBotoClientWrapper(conn, retry_decorator)
2022-02-01 14:55:16.224186 | controller | E       AttributeError: module 'ansible_collections.amazon.aws.plugins.module_utils.core' has no attribute '_RetryingBotoClientWrapper'
2022-02-01 14:55:16.224190 | controller |
2022-02-01 14:55:16.224192 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py:86: AttributeError
2022-02-01 14:55:16.224195 | controller | __________________ test_find_connection_insufficient_filters ___________________
2022-02-01 14:55:16.224198 | controller | [gw0] linux -- Python 3.8.12 /home/zuul/venv/bin/python
2022-02-01 14:55:16.224202 | controller | placeboify = Session(region_name='us-west-2'), maybe_sleep = None
2022-02-01 14:55:16.224204 | controller |
2022-02-01 14:55:16.224207 | controller |     def test_find_connection_insufficient_filters(placeboify, maybe_sleep):
2022-02-01 14:55:16.224211 | controller |         # get list of customer gateways and virtual private gateways
2022-02-01 14:55:16.224213 | controller |         cgw, vgw = get_dependencies()
2022-02-01 14:55:16.224217 | controller |
2022-02-01 14:55:16.224220 | controller |         # create two connections with the same tags
2022-02-01 14:55:16.224223 | controller |         params = make_params(cgw[0], vgw[0], tags={'Correct': 'Tag'})
2022-02-01 14:55:16.224225 | controller |         params2 = make_params(cgw[1], vgw[1], tags={'Correct': 'Tag'})
2022-02-01 14:55:16.224228 | controller | >       m, conn = setup_mod_conn(placeboify, params)
2022-02-01 14:55:16.224231 | controller |
2022-02-01 14:55:16.224238 | controller | tests/unit/plugins/modules/test_ec2_vpc_vpn.py:171:
```

I don't think this is related to `ec2_scaling_policy`.

**Comment by falcon78921** [comment]
@jillr The sanity tests finally passed! :) When running the AWS integration tests in my environment, everything checks out ok. Not sure if I'm missing something.

**Comment by falcon78921** [comment]
Hey @tremble :wave:

When you have a chance, could you please take a peek at this? Not sure if I'm missing something on the integration side.

**Comment by falcon78921** [comment]
Hey @alinabuzachis :wave:

If you have a chance, could you please take a look at this PR and see if anything needs fixed? The sanity tests all pass now. The only roadblock seems to be the AWS integration tests. The same test suite works when I run it in my environment. Maybe I'm missing something at the CI/CD level?

**Comment by falcon78921** [comment]
@tremble @alinabuzachis Just following up on a review for this PR

**Comment by alinabuzachis** [comment]
@falcon78921 Sorry for the delay. Will review it later on today and let you know.

File: plugins/modules/ec2_scaling_policy.py [hunk 7]
```diff
@@ -107,6 +110,79 @@
         description:
           - The amount by which to scale.
         required: true
+  target_tracking_config:
```
Comment (alinabuzachis) [inline plugins/modules/ec2_scaling_policy.py:113]: add `version_added: 3.2.0`

**Review by alinabuzachis — APPROVED** [review]
@falcon78921  This LGTM, just add `version_added`. Thank you.

File: plugins/modules/ec2_scaling_policy.py (hunk 7 above)
Comment (falcon78921) [reply plugins/modules/ec2_scaling_policy.py:113]: added a `version_added` key/value pair. thanks @alinabuzachis

**Comment by falcon78921** [comment]
Hey @markuman :wave:, just following up on this PR and seeing if there's anything else I need to do.

**Comment by markuman** [comment]
recheck

**Comment by falcon78921** [comment]
(from the latest sanity check)
```
Assuming Docker is available on localhost.
Run command: ssh-keygen -m PEM -q -t rsa -N '' -f /home/zuul/.ansible/test/id_rsa
Run command: podman -v
Detected "podman" container runtime version: podman version 3.4.4
Starting new "ansible-test-controller-WjNFpBFY" container.
Run command: podman system connection list --format=json
ERROR: Host <ansible_test._internal.host_profiles.DockerProfile object at 0x7f4dbe2cb490> job failed: Expecting value: line 1 column 1 (char 0)
```

**Comment by markuman** [comment]
@falcon78921 can you rebase your PR?

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
recheck

**Comment by tremble** [comment]
recheck

**Pull request merged**